os/bluestore: introduce allocator state histogram

ifed01 opened the pull request:
And access it through admin socket

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -17764,9 +17764,11 @@ void BlueStore::_log_alerts(osd_alert_list_t& alerts)
-  alloc_stats_count++;
-  alloc_stats_fragments += extents.size();
-  alloc_stats_size += need;
+  if (alloc_size != min_alloc_size) {
```
I do not understand the exclusion that has been make here.
Even for smaller allocation we we still did make an allocation; and it has size.
Granted, the count of fragments will be 1.
How is counting that in obfuscating the allocation report?

ifed01 replied:
The primary purpose of these allocation stats/probes is to get some overview on how fragmented allocations get over time. Single unit allocations make no sense for this purpose and just spoil these stats. 
Hence removed.

aclamk commented on src/os/bluestore/Allocator.cc:
```diff
@@ -100,6 +107,57 @@ class Allocator::SocketHook : public AdminSocketHook {
+        } else if (len > max) {
+          idx = num_buckets - 1;
+        } else {
+          size_t most_bit = cbits((uint64_t(len))) - 1;
```
NIT: 
most_bit = cbits(len - 1); //len will not be 0, ever
//11..0  -> idx 0
//13..12 -> idx 1
//15..14 -> idx 2
idx = (most_bit - 10) / 2;

ifed01 replied:
This required some adjustment since I'd like to include 16K, 64K,.. sizes to previous bucket so we get ranges like that: [0..4K] (4K..16K] (16K..64K] while you formula results in [4K..16K) [16K..64K) ranges..
Anyway thanks for the hint!

aclamk commented on src/os/bluestore/Allocator.cc:
```diff
@@ -100,6 +107,57 @@ class Allocator::SocketHook : public AdminSocketHook {
+        ++counts[idx];
+      };
+      alloc->foreach(iterated_allocation);
+      f->open_array_section("extent counts");
```
I think " " here makes json parser unhappy.

ifed01 replied:
Fixed

aclamk reviewed (commented):
Good addition.

aclamk commented:
The new and improvement reporting looks like this (xml-pretty):
```
<extent_counts>
...
 <c>
  <max_len>65536</max_len>
  <total>2</total>
  <aligned>2</aligned>
  <units>31</units>
 </c>
```
I think that report should
1) include "min_len"
2) "total"->"count"
3) "aligned"-> I do not think this value is useful at all
4) "units" -> "usable" + "unusable", and I think it should be bytes, not AUs
5) "c" -> "bin" ?

ifed01 commented:
> The new and improvement reporting looks like this (xml-pretty):
> 
> ```
> <extent_counts>
> ...
>  <c>
>   <max_len>65536</max_len>
>   <total>2</total>
>   <aligned>2</aligned>
>   <units>31</units>
>  </c>
> ```
> 
> I think that report should
> 
> 1. include "min_len"
done
> 2. "total"->"count"
renamed to extent_count
> 3. "aligned"-> I do not think this value is useful at all
renamed to aligned_extent_count. Let it be. 'usable_bytes' field below is of more use though.
> 4. "units" -> "usable" + "unusable", and I think it should be bytes, not AUs
usable_bytes & total_bytes created.

> 5. "c" -> "bin" ?
done

ifed01 commented:
jenkins test api

ifed01 commented:
jenkins test api

ifed01 commented:
jenkins test api

ifed01 commented:
jenkins test make check

ifed01 commented:
jenkins test api

yuriw commented:
jenkins test make check

ifed01 commented:
jenkins test make check

yuriw commented:
jenkins test make check

yuriw commented:
jenkins test make check

ifed01 commented:
jenkins test windows
